File: 1563-stone-game-v/1563-stone-game-v.py
from typing import List
from functools import cache
from itertools import accumulate


class Solution:
    def stoneGameV(self, stoneValue: List[int]) -> int:

        prefix = list(accumulate(stoneValue, initial=0))

        @cache
        def dfs(left: int, right: int) -> int:
            if left == right:
                return 0

            ans = 0
            left_sum = 0

            total = prefix[right + 1] - prefix[left]

            for k in range(left, right):
                left_sum += stoneValue[k]
                right_sum = total - left_sum

                # Left part is smaller
                if left_sum < right_sum:

                    # Even the maximum possible result
                    # cannot beat our current answer.
                    if ans >= 2 * left_sum:
                        continue

                    ans = max(
                        ans,
                        left_sum + dfs(left, k)
                    )

                # Right part is smaller
                elif left_sum > right_sum:

                    # As k increases, right_sum only decreases.
                    # So no future split can improve the answer.
                    if ans >= 2 * right_sum:
                        break

                    ans = max(
                        ans,
                        right_sum + dfs(k + 1, right)
                    )

                # Equal sums
                else:
                    ans = max(
                        ans,
                        left_sum + dfs(left, k),
                        right_sum + dfs(k + 1, right)
                    )

            return ans

        return dfs(0, len(stoneValue) - 1)

File: 1563-stone-game-v/1563-stone-game-v.c
#include <stdlib.h>

int stoneGameV(int* stoneValue, int stoneValueSize) {
    int n = stoneValueSize;

    long long* prefix =
        (long long*)malloc((n + 1) * sizeof(long long));

    for (int i = 0; i <= n; i++) {
        prefix[i] = 0;
    }

    for (int i = 0; i < n; i++) {
        prefix[i + 1] = prefix[i] + stoneValue[i];
    }

    int** dp = (int**)malloc(n * sizeof(int*));

    for (int i = 0; i < n; i++) {
        dp[i] = (int*)calloc(n, sizeof(int));
    }

    for (int len = 2; len <= n; len++) {
        for (int l = 0; l + len <= n; l++) {
            int r = l + len - 1;

            for (int k = l; k < r; k++) {
                long long leftSum =
                    prefix[k + 1] - prefix[l];

                long long rightSum =
                    prefix[r + 1] - prefix[k + 1];

                if (leftSum < rightSum) {
                    int value =
                        (int)leftSum + dp[l][k];

                    if (value > dp[l][r])
                        dp[l][r] = value;
                }
                else if (leftSum > rightSum) {
                    int value =
                        (int)rightSum + dp[k + 1][r];

                    if (value > dp[l][r])
                        dp[l][r] = value;
                }
                else {
                    int best =
                        dp[l][k] > dp[k + 1][r]
                        ? dp[l][k]
                        : dp[k + 1][r];

                    int value =
                        (int)leftSum + best;

                    if (value > dp[l][r])
                        dp[l][r] = value;
                }
            }
        }
    }

    int answer = dp[0][n - 1];

    for (int i = 0; i < n; i++) {
        free(dp[i]);
    }

    free(dp);
    free(prefix);

    return answer;
}